enhancements/ingress/aws-load-balancer-operator.md

```diff
@@ -0,0 +1,394 @@
+    route requests based on the HTTP path and host. It also integrates with other
+    AWS services like DDoS protection and WAF.
+
+### Openshift Router
+In OCP/OKD Ingress resources are serviced by the Openshift
+[router](https://github.com/openshift/router). Routers are created by the
+[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/)(CIO)
+for every
+_[IngressController](https://docs.openshift.com/container-platform/4.9/networking/configuring_ingress_cluster_traffic/configuring-ingress-cluster-traffic-ingress-controller.html)_
+resource. By default on AWS the CIO creates a Service of type
+[LoadBalancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer).
+The in-tree Service controller in turn provisions a Network Load
+Balancer with a TCP listener. The CIO then creates a DNS entry for the subdomain
+specified in the IngressController resource with the load balancer created in
+the previous step as the target. 
+
+### aws-load-balancer-controller
+AWS maintains an out-of-tree controller with support for ALBs for Ingress and
+Service type resources. The in-tree Service controller only provisions [classic
+load
+balancers](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/introduction.html)
+and
+[Network](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/introduction.html)
+Load Balancers. The out-of-tree controller supports integration of Kubernetes
+resources with more of the AWS services like WAFs and the DDoS prevent service
+[Shield](https://docs.aws.amazon.com/waf/latest/developerguide/ddos-overview.html).
+It also enables traffic routing from the load balancer directly to the pods of
+an Ingress or Service if the cluster uses the [AWS VPC Container Networking
+Interface](https://github.com/aws/amazon-vpc-cni-k8s). In addition it also
+introduces 2 Custom Resources namely
+[`IngressClassParams`](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.3/guide/ingress/ingress_class/#ingressclassparams)
+and
+[`TargetGroupBindings`](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.3/guide/targetgroupbinding/targetgroupbinding/).
+_IngressClassParams_ is used for specifying custom parameters for Ingresses of a
+certain class.  The controller uses _TargetGroupBinding_ internally and it can
+also be used by users to manage the load balancer externally. The
+aws-load-balancer will henceforth be referred to as **“lb-controller”** for the
+sake of brevity.
+
+## Motivation
+
+Implement an operator, tentatively called `aws-load-balancer-operator`, which
+deploys and manages an instance of the lb-operator. The operator will be
+distributed through Operator Hub.
+
+### Goals
+* Implement operator which can be installed and used in OCP/OKD.
+
+### Non-Goals
+* Implementing support for ALB for Openshift Routes
+* Support for multiple installations of the controller i.e. multiple custom
+  resources.
```

Comment:
Is the API being designed to accommodate a future enhancement adding this capability in case it should be come possible upstream and users should request the capability?  

Actually, I want to ask this question about the other non-goals too, but especially this one.  

On further thought and on reading https://github.com/openshift/enhancements/pull/1020/files#r794420487, does it make sense at all to have multiple instances running?  It seems that every instance will match every ingress that has an ingressclass that has `spec.controller: ingress.k8s.aws/alb`.  I suppose, though, that having multiple instances could be useful in the future if we also added support for managing services or routes, right?

Reply:
> Is the API being designed to accommodate a future enhancement adding this capability in case it should be come possible upstream and users should request the capability?
Yes, when this becomes possible upstream users will be able to create multiple controllers. Right now we only reconcile one resourcce called `cluster`.

Reply:
> does it make sense at all to have multiple instances running?

We don't restrict by namespace in the current implementation but the controller does support it. So I suppose when the multiple instance support does come we will have to support the namespace configuration as well. But we will have to ensure that there is no situation where multiple instances of the controller don't all try to reconcile the same Ingress resource due to overlap.

Reply:
This is just speculation but when multi-instance support does come it may be possible to have different instances reconciling Ingresses of the same Ingress class in different namespaces. We will have to be careful in how we allow the user to configure the instances so that there is no contention between the instances.